File: frontend-mobile/biblio_upb/lib/models/emprunt.dart
class Emprunt {
  final int? empruntId;
  final String dateEmprunt;
  final String dateRetourPrevue;
  final String? dateRetourEffective;
  final String statut;
  final int nombreProlongations;
  final Map<String, dynamic>? livre;

  Emprunt({
    this.empruntId,
    required this.dateEmprunt,
    required this.dateRetourPrevue,
    this.dateRetourEffective,
    required this.statut,
    this.nombreProlongations = 0,
    this.livre,
  });

  factory Emprunt.fromJson(Map<String, dynamic> json) => Emprunt(
    empruntId: json['empruntId'],
    dateEmprunt: json['dateEmprunt'] ?? '',
    dateRetourPrevue: json['dateRetourPrevue'] ?? '',
    dateRetourEffective: json['dateRetourEffective'],
    statut: json['statut'] ?? 'EN_COURS',
    nombreProlongations: json['nombreProlongations'] ?? 0,
    livre: json['livre'],
  );

  String get titreLivre => livre?['titre'] ?? 'Livre inconnu';
  bool get enRetard => statut == 'EN_RETARD';
  bool get peutProlonger => nombreProlongations == 0;
}

File: backend/bibliotheque/src/main/java/com/upb/bibliotheque/entity/Emprunt.java
package com.upb.bibliotheque.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDate;

@Entity
@Table(name = "emprunts")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor
public class Emprunt {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long empruntId;

    @ManyToOne
    @JoinColumn(name = "etudiant_id", nullable = false)
    private Etudiant etudiant;

    @ManyToOne
    @JoinColumn(name = "livre_id", nullable = false)
    private Livre livre;

    @Column(nullable = false)
    private LocalDate dateEmprunt = LocalDate.now();
    @Column(nullable = false)
    private LocalDate dateRetourPrevue = LocalDate.now().plusDays(14);
    private LocalDate dateRetourEffective;

    @Enumerated(EnumType.STRING)
    private StatutEmprunt statut = StatutEmprunt.EN_COURS;
    private Integer nombreProlongations = 0;

    public enum StatutEmprunt { EN_COURS, RETOURNE, EN_RETARD }
    public boolean estEnRetard() { return dateRetourEffective == null && LocalDate.now().isAfter(dateRetourPrevue); }
}